Replace startup code block with loop to call all entry stubs.

tmcgilchrist opened the pull request:
This is a small improvement to rewrite the entry_point code generated by `cmm_helpers.ml` to be data driven.

When starting a native OCaml program the call stack goes through C functions (caml_main, caml_startup_common and caml_startup_common,  in `startup_nat.c` before jumping to the processor specific assembly in `runtime/arm64.s` (or whatever architecture you have) starting at `caml_start_program`. Before it jumps to `caml_program` which calls the module `.entry` point for the modules included in an executable. At this point currently OCaml generates a block of assembly code for each module entry point that needs to be called, 35 for a minimal program that just relies on the Stdlib (potentially more?). This change re-writes this code to be a loop equivalent to this C code:

``` c
 {
   int id = 0;
   while (true) {
     if (id == len_caml_globals_entry_functions) goto out;
     caml_globals_entry_functions[id]();
     caml_globals_inited += 1;
     id += 1;
   }
   out:
   return 1;
 }
```

What is gained by making this change? More compact assembly code that is data driven using a loop and lookup table of symbols. Possibly faster startup but I have no data to back this up. The work was done original in PR https://github.com/ocaml-flambda/flambda-backend/pull/1425 by @poechsel, I have only cherry picked it here after noticing the repeated assembly code blocks.

@poechsel please correct any attribution as you see fit.

tmcgilchrist committed:
Replace startup code block with loop to call all entry stubs.

tmcgilchrist committed:
Generate caml_program as a loop over module entry points

Previously caml_program was emitted as a loop unrolled once per linked
module, with each module's entry stub called in turn followed by an
increment of caml_globals_inited. When linking large binaries,
generating this function accounted for a sizable share of link time,
most of it spent in register allocation and assembling the ever-growing
body.

Instead, emit the module entry-point addresses as a table
(caml_globals_entry_functions) in .data and generate caml_program
as a loop that walks the table calling each entry. The size of the
startup code is now independent of the number of linked modules.

Originally implemented in OxCaml by Pierre Oechsel
(https://github.com/oxcaml/oxcaml/pull/1425).

Co-authored-by: Pierre Oechsel <pierre.oechsel@gmail.com>

shindere commented:
Thanks a lot for the nice PR description!

Out of curiosity: did you check that the change also works when frame
pointers are enabled?

tmcgilchrist commented:
Good idea @shindere I have not as yet, but I will do so.

shindere commented:
Tim McGilchrist (2024/06/05 00:18 -0700):
> Good idea @shindere I have not as yet, but I will do so.

Thanks! I'll wait for you to post here to proceed, although I am likely
not the most well suited to review this.

tmcgilchrist commented:
@shindere testing with frame-pointers on macOS and Linux is passing.

shindere commented:
Tim McGilchrist (2024/06/18 08:08 -0700):
> @shindere testing with frame-pointers on macOS and Linux is passing.

Okay, thanks!

smuenzel commented:
@tmcgilchrist, are you still interested in getting this merged? If so, I could work on reviewing it.

tmcgilchrist commented:
Thank you @smuenzel that would be appreciated.

smuenzel commented on asmcomp/cmm_helpers.ml:
```diff
@@ -2504,34 +2504,107 @@ let emit_float_array_constant symb fields cont =
+              dbg,
+            Ctuple [],
+            dbg))
+  in
```
It might be nice to avoid the mutable variable, leading to shorter cmm code, but identical assembly code:
```ocaml
  let body =
    let dbg = dbg () in
    let repeat_raise = Lambda.next_raise_count () in
    Ccatch
      ( Recursive,
        [ repeat_raise
        , [ id, typ_int ]
        , Cifthenelse
            ( Cop (Ccmpi Ceq, [var_id; high], dbg),
              dbg,
              Ctuple [],
              dbg,
              Csequence
                ( call (var_id)
                , Cexit (repeat_raise, [ Cop (Caddi, [ var_id; Cconst_int (1, dbg)], dbg)])
                ),
              dbg
            )
        , dbg
        ],
        Cexit (repeat_raise, [ Cconst_int (0,dbg)])
      )
  in
```

tmcgilchrist replied:
Looks good, I've included this change.

smuenzel reviewed (approved):
I took a look and this PR does exactly what it says. When I tested it on trunk, the ocamlopt.opt executable was one block shorter (4KB). Obviously not a huge gain (0.021%), but the result seems cleaner.

I have a small suggestion regarding shortening the Cmm code (which has no effect on the outputted assembly), but overall, I think this is good to go.

smuenzel reviewed (approved):
Looks good!

@gasche, sorry to bother you with this, but if you think my review is sufficient, can you give this the green check?

gasche reviewed (approved):
Approved on @smuenzel's behalf (I also looked at this and it looks reasonable).

gasche commented:
I would prefer if the commit was attributed to its actual author (which I understand to be @poechsel). (Sometimes the commit gets reword substantially on the way and a `Co-authored-by: ...` is fine.) @tmcgilchrist do you think it would be okay to change the author before merging?

I think it would also be nice if the commit message pointed to the oxcaml PR and/or commit number, so that we can more easily find the original PR and discussion from the git history.

tmcgilchrist commented:
@gasche I've modified the commit message to link back to the OxCaml PR and added `Co-authored-by:`, I think there's enough rework done to choose that option.

dra27-js commented:
I think in general we should avoid synthetically editing authors on commits - it's a bad habit. For things coming from OxCaml, I suggest exercising judgement - if it's reasonable to start from a cherry-pick, use `git cherry-pick -x` which will "preserve" the original author on the oxcaml/oxcaml commit and reference the commit. If, on the other hand, there's considerable editing and it's easier to adapt the changes from scratch, I think adding `Co-authored-by` (as here) is fine.

gasche commented:
I am fine with Co-authored-by, (I think that when two separate people work on a PR, it is also reasonable to squash a bit less: keep the original commit unchanged, ideally with `cherry-pick -x` indeed, and then add additional commits by the submitter on top to reflect further adaptation/review/cleanup work.)
